### PR TITLE
Support port collections in GRC YAML   

### PR DESCRIPTION
**Pull request by frankosterfeld**
Make sure we can load/save connections involving port collections, i.e. serialize a connection from block1, port collection 0, sub-index 1 to block2, port collection 0, sub-index 2 as
    
```
 connections:
   - [block1, [0, 1], block2, [0, 2]]
```

### CODE REVIEW

File: core/include/gnuradio-4.0/Graph.hpp [hunk 1]
```diff
@@ -431,39 +431,48 @@ class BlockWrapper : public BlockModel {
     }
 };
 
-class edge {
+template<typename T>
+struct PortIndexDefinition {
+    T           topLevel;
```
Comment (RalphSteinhagen) [inline core/include/gnuradio-4.0/Graph.hpp:436]: If you could change `T` -> `std::size_t` and remove the unnecessary templatisation and constructor.

File: core/test/app_grc.cpp [hunk 2]
```diff
@@ -13,13 +13,45 @@
 
 namespace grg = gr;
 
+template<typename T>
+struct ArraySource : public grg::Block<ArraySource<T>> {
```
Comment (RalphSteinhagen) [inline core/test/app_grc.cpp:17]: boy-scout-rule: could you also rename `grg` -> `gr`

File: core/test/app_grc.cpp [hunk 3]
```diff
@@ -13,13 +13,45 @@
...
+
+template<typename T>
+struct ArraySink : public grg::Block<ArraySink<T>> {
+    std::array<grg::PortIn<T>, 2> inA;
+    std::array<grg::PortIn<T>, 2> inB;
+};
+
+ENABLE_REFLECTION_FOR_TEMPLATE(ArraySink, inA, inB);
+
 struct test_context {
```
Comment (RalphSteinhagen) [inline core/test/app_grc.cpp:32]: boy-scout-rule: if you could refactor and change `test_context` -> `TestContext`

**Review by RalphSteinhagen — APPROVED** [review]
LGTM! :+1: 

Some minor 'boy-scout-rule' things. If you could address them and then merge. You have my OK.

File: core/test/app_grc.cpp (hunk 2 above)
Comment (frankosterfeld) [reply core/test/app_grc.cpp]: Done

File: core/test/app_grc.cpp (hunk 3 above)
Comment (frankosterfeld) [reply core/test/app_grc.cpp]: Done

File: core/include/gnuradio-4.0/Graph.hpp (hunk 1 above)
Comment (frankosterfeld) [reply core/include/gnuradio-4.0/Graph.hpp:436]: PortIndexDefinition is used with T=std::size_t and T=std::string, as an argument to the connect() calls, so it's desirable from an API POV that the passed string/std::size_t parameters are implicitly converted to PortIndexDefinition. (thus the non-explicit ctor)

Comment (RalphSteinhagen) [reply core/include/gnuradio-4.0/Graph.hpp:436]: OK.

**Pull request merged**